File: packages/companion/swift/Sources/BridgethingCompanion/AudioDispatcher.swift
import BridgethingGateway
import BridgethingGlue
import BridgethingSchema
import Foundation

public actor AudioDispatcher {
    private let backend: any AudioBackend
    private var tasks: [Task<Void, Never>] = []
    private var glueProvider: (@Sendable () async -> (any BridgethingGlue)?)?

    public init(backend: any AudioBackend) {
        self.backend = backend
    }

    public func setGlueProvider(_ provider: @escaping @Sendable () async -> (any BridgethingGlue)?) {
        glueProvider = provider
    }

    private func volumeGlue() async -> (any BridgethingGlue)? {
        guard let glue = await glueProvider?(), await glue.ownsVolume() else { return nil }
        return glue
    }

    public func start(gateway: BridgethingGateway) async {
        let backend = backend
        tasks.append(Task { [weak self] in
            for await _ in gateway.audio.volumeUp {
                guard let glue = await self?.volumeGlue() else { await backend.volumeUp(); continue }
                do { try await glue.volumeUp() } catch { await Self.reportRejected("volumeUp", error, gateway: gateway) }
            }
        })
        tasks.append(Task { [weak self] in
            for await _ in gateway.audio.volumeDown {
                guard let glue = await self?.volumeGlue() else { await backend.volumeDown(); continue }
                do { try await glue.volumeDown() } catch { await Self.reportRejected("volumeDown", error, gateway: gateway) }
            }
        })
        tasks.append(Task { [weak self] in
            for await (_, msg) in gateway.audio.setVolume {
                guard let glue = await self?.volumeGlue() else { await backend.setVolume(msg.level); continue }
                do { try await glue.setVolume(msg.level) } catch { await Self.reportRejected("setVolume", error, gateway: gateway) }
            }
        })
        tasks.append(Task { [weak self] in
            for await _ in gateway.audio.muteToggle {
                if await self?.volumeGlue() == nil { await backend.muteToggle() }
            }
        })
        tasks.append(Task { [weak self] in
            for await (_, msg) in gateway.audio.setMute {
                if await self?.volumeGlue() == nil { await backend.setMute(msg.muted) }
            }
        })
        tasks.append(Task { for await (_, msg) in gateway.audio.ttsCancel { await backend.cancel(id: msg.id) } })
        tasks.append(Task { for await _ in gateway.audio.ttsCancelAll { await backend.cancelAll() } })
        tasks.append(Task {
            for await (_, msg) in gateway.audio.earcon where await !backend.playEarcon(name: msg.name) {
                await Self.reportAudioError(
                    .unavailable(AudioErrorUnavailableInner(verb: "earcon")),
                    gateway: gateway
                )
            }
        })
        tasks.append(Task { [weak self] in
            for await (_, msg) in gateway.audio.tts {
                await self?.handleTts(msg, gateway: gateway)
            }
        })
    }

    public func stop() async {
        for task in tasks { task.cancel() }
        tasks.removeAll()
        await backend.cancelAll()
    }

    private static func reportRejected(_ verb: String, _ error: Error, gateway: BridgethingGateway) async {
        await reportAudioError(
            .actionRejected(AudioErrorActionRejectedInner(reason: "\(verb): \(error.localizedDescription)")),
            gateway: gateway
        )
    }

    private static func reportAudioError(_ error: AudioError, gateway: BridgethingGateway) async {
        try? await gateway.audio.errorEvent(AudioErrorReply(error: error))
    }

    private func handleTts(_ msg: Tts, gateway: BridgethingGateway) async {
        let id = msg.id
        let completed = await backend.speak(id: id, text: msg.text, voice: msg.voice) {
            Task { try? await gateway.audio.ttsStarted(TtsStarted(id: id)) }
        }
        try? await gateway.audio.ttsEnded(TtsEnded(id: id, completed: completed))
    }
}

File: packages/companion/kotlin/companion/src/main/kotlin/com/bridgething/companion/AudioDispatcher.kt
package com.bridgething.companion

import com.bridgething.gateway.BridgethingGateway
import com.bridgething.gateway.audio
import com.bridgething.schema.AudioError
import com.bridgething.schema.AudioErrorActionRejectedInner
import com.bridgething.schema.AudioErrorReply
import com.bridgething.schema.AudioErrorUnavailableInner
import com.bridgething.glue.BridgethingGlue
import com.bridgething.schema.Tts
import com.bridgething.schema.TtsEnded
import com.bridgething.schema.TtsStarted
import com.bridgething.schema.VolumeChanged
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

public class AudioDispatcher(
    private val backend: AudioBackend,
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
    private val mutex = Mutex()
    private val jobs = mutableListOf<Job>()

    @Volatile
    private var glueProvider: (suspend () -> BridgethingGlue?)? = null

    public fun setGlueProvider(provider: suspend () -> BridgethingGlue?) {
        glueProvider = provider
    }

    private suspend fun volumeGlue(): BridgethingGlue? {
        val glue = glueProvider?.invoke() ?: return null
        return if (glue.ownsVolume()) glue else null
    }

    public suspend fun start(gateway: BridgethingGateway) {
        mutex.withLock {
            stopJobs()
            jobs.add(
                scope.launch {
                    gateway.audio.volumeUp.collect {
                        val glue = volumeGlue()
                        if (glue != null) {
                            runCatching { glue.volumeUp() }
                                .onSuccess { level -> broadcastVolume(gateway, level) }
                                .onFailure { reportRejected(gateway, "volumeUp", it) }
                        } else {
                            backend.volumeUp()
                        }
                    }
                },
            )
            jobs.add(
                scope.launch {
                    gateway.audio.volumeDown.collect {
                        val glue = volumeGlue()
                        if (glue != null) {
                            runCatching { glue.volumeDown() }
                                .onSuccess { level -> broadcastVolume(gateway, level) }
                                .onFailure { reportRejected(gateway, "volumeDown", it) }
                        } else {
                            backend.volumeDown()
                        }
                    }
                },
            )
            jobs.add(
                scope.launch {
                    gateway.audio.setVolume.collect { (_, msg) ->
                        val glue = volumeGlue()
                        if (glue != null) {
                            runCatching { glue.setVolume(msg.level) }
                                .onSuccess { level -> broadcastVolume(gateway, level) }
                                .onFailure { reportRejected(gateway, "setVolume", it) }
                        } else {
                            backend.setVolume(msg.level)
                        }
                    }
                },
            )
            jobs.add(
                scope.launch {
                    gateway.audio.muteToggle.collect {
                        if (volumeGlue() == null) backend.muteToggle()
                    }
                },
            )
            jobs.add(
                scope.launch {
                    gateway.audio.setMute.collect { (_, msg) ->
                        if (volumeGlue() == null) backend.setMute(msg.muted)
                    }
                },
            )
            jobs.add(scope.launch { gateway.audio.ttsCancel.collect { (_, msg) -> backend.cancel(msg.id) } })
            jobs.add(scope.launch { gateway.audio.ttsCancelAll.collect { backend.cancelAll() } })
            jobs.add(
                scope.launch {
                    gateway.audio.earcon.collect { (_, msg) ->
                        if (!backend.playEarcon(msg.name)) {
                            reportAudioError(gateway, AudioError.Unavailable(AudioErrorUnavailableInner("earcon")))
                        }
                    }
                },
            )
            jobs.add(scope.launch { gateway.audio.tts.collect { (_, msg) -> handleTts(msg, gateway) } })
        }
    }

    private suspend fun reportRejected(gateway: BridgethingGateway, verb: String, error: Throwable) {
        reportAudioError(
            gateway,
            AudioError.ActionRejected(AudioErrorActionRejectedInner("$verb: ${error.message ?: error.toString()}")),
        )
    }

    private suspend fun reportAudioError(gateway: BridgethingGateway, error: AudioError) {
        runCatching { gateway.audio.errorEvent(AudioErrorReply(error)) }
    }

    public suspend fun stop() {
        mutex.withLock { stopJobs() }
        runCatching { backend.cancelAll() }
    }

    public fun close() {
        scope.cancel()
    }

    private fun stopJobs() {
        for (job in jobs) job.cancel()
        jobs.clear()
    }

    private suspend fun broadcastVolume(gateway: BridgethingGateway, level: Float) {
        runCatching { gateway.audio.volumeChanged(VolumeChanged(level = level, muted = false)) }
    }

    private suspend fun handleTts(msg: Tts, gateway: BridgethingGateway) {
        val id = msg.id
        val completed = backend.speak(id, msg.text, msg.voice) {
            scope.launch { runCatching { gateway.audio.ttsStarted(TtsStarted(id)) } }
        }
        runCatching { gateway.audio.ttsEnded(TtsEnded(id, completed)) }
    }
}
